src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,274 @@
+	r.RLock()
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.RUnlock()
+
+	if r.closed || r.pastLastShard {
+		return size, totalDocsCount, ErrWideQueryResultsExhausted
+	}
+
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.Unlock()
+
+	if err != nil && err != ErrWideQueryResultsExhausted {
+		// NB: if exhausted, drain the current batch and overflows.
+		return size, totalDocsCount, err
+	}
+
+	release := len(r.batch.IDs) == r.batchSize
+	if release {
+		r.releaseAndWait()
+		r.releaseOverflow()
+	}
+
+	r.RLock()
+	size = r.size
+	r.RUnlock()
+
+	return size, totalDocsCount, err
+}
+
+func (r *wideResults) addDocumentsBatchWithLock(batch []doc.Document) error {
+	for i := range batch {
+		if err := r.addDocumentWithLock(batch[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *wideResults) addDocumentWithLock(d doc.Document) error {
+	if len(d.ID) == 0 {
+		return errUnableToAddResultMissingID
+	}
+
+	var tsID ident.ID = ident.BytesID(d.ID)
+
+	// Need to apply filter if set first.
+	if r.shardFilter != nil {
+		filteringShard := r.shards[r.shardIdx]
+		documentShard, documentShardOwned := r.shardFilter(tsID)
+		// NB: Check to see if shard is exceeded first (to short circuit earlier if
+		// the current shard is not owned by this node, but shard exceeds filtered).
+		if filteringShard > documentShard {
+			// this document is from a shard lower than the next shard allowed.
+			return nil
+		}
+
+		for filteringShard < documentShard {
+			// this document is from a shard higher than the next shard allowed;
+			// advance to the next shard, then try again.
+			r.shardIdx++
+			if r.shardIdx >= len(r.shards) {
+				// shard is past the final shard allowed by filter, no more results
+				// will be accepted.
+				r.pastLastShard = true
+				return ErrWideQueryResultsExhausted
+			}
+
+			filteringShard = r.shards[r.shardIdx]
+			if filteringShard > documentShard {
+				// this document is from a shard lower than the next shard allowed.
+				return nil
+			}
+		}
+
+		if !documentShardOwned {
+			return nil
+		}
+	}
+
+	r.size++
+	r.totalDocsCount++
+
+	// Pool IDs after filter is passed.
+	tsID = r.idPool.Clone(tsID)
+	if len(r.batch.IDs) < r.batchSize {
+		r.batch.IDs = append(r.batch.IDs, tsID)
+	} else {
+		r.idsOverflow = append(r.idsOverflow, tsID)
+	}
+
+	return nil
+}
+
+func (r *wideResults) Namespace() ident.ID {
+	return r.nsID
+}
+
+func (r *wideResults) Size() int {
+	r.RLock()
+	v := r.size
+	r.RUnlock()
+	return v
+}
+
+func (r *wideResults) TotalDocsCount() int {
+	r.RLock()
+	v := r.totalDocsCount
+	r.RUnlock()
+	return v
+}
+
+// NB: Finalize should be called after all documents have been consumed.
+func (r *wideResults) Finalize() {
+	if r.closed {
+		return
+	}
+
+	// NB: release current
+	r.releaseAndWait()
+	r.closed = true
+	close(r.batchCh)
+}
+
+func (r *wideResults) releaseAndWait() {
+	if r.closed || len(r.batch.IDs) == 0 {
```

Comment:
Seems everything should be done in a lock here no? Lots of access of fields that could be concurrently being read/written I think?

Reply:
Yeah will shuffle some things around, think i can make it a little clearer too